cmds/vbe.c

```diff
@@ -0,0 +1,349 @@
+	graphmode_t fallback = { 0 };
+
+	for (i = 0; modeIds[i] != 0xffff; i++) {
+		err = _vbe_modeInfoGet(modeIds[i]);
+		if (err < 0) {
+			return -1;
+		}
+
+		if ((vbe_common.modeInfo->ModeAttributes & SUPPORTS_LFB) != 0) {
+			if ((vbe_common.modeInfo->XResolution == vbe_common.preferred.width) && (vbe_common.modeInfo->YResolution == vbe_common.preferred.height) && (vbe_common.modeInfo->BitsPerPixel == vbe_common.preferred.bpp)) {
+				picked->width = vbe_common.modeInfo->XResolution;
+				picked->height = vbe_common.modeInfo->YResolution;
+				picked->bpp = vbe_common.modeInfo->BitsPerPixel;
+				picked->pitch = vbe_common.modeInfo->BytesPerScanLine;
+				picked->framebuffer = vbe_common.modeInfo->PhysBasePtr;
+
+				return modeIds[i];
+			}
+
+			if ((vbe_common.modeInfo->XResolution == vbe_common.fallback.width) && (vbe_common.modeInfo->YResolution == vbe_common.fallback.height) && (vbe_common.modeInfo->BitsPerPixel == vbe_common.fallback.bpp)) {
+				fallbackModeId = modeIds[i];
+				fallback.width = vbe_common.modeInfo->XResolution;
+				fallback.height = vbe_common.modeInfo->YResolution;
+				fallback.bpp = vbe_common.modeInfo->BitsPerPixel;
+				fallback.pitch = vbe_common.modeInfo->BytesPerScanLine;
+				fallback.framebuffer = vbe_common.modeInfo->PhysBasePtr;
+			}
+		}
+	}
+
+	hal_memcpy(picked, &fallback, sizeof(graphmode_t));
+	return fallbackModeId;
+}
+
+
+static int vbe_videoModeParse(char *str, video_mode_t *video_mode)
+{
+	unsigned int val;
+	char *ptr = str;
+
+	if ((ptr == NULL) || (*ptr == '\0')) {
+		return -EINVAL;
+	}
+
+	val = lib_strtoul(ptr, &ptr, 0);
+	if (*ptr != 'x') {
+		return -EINVAL;
+	}
+	video_mode->width = val;
+
+	ptr++;
+	val = lib_strtoul(ptr, &ptr, 0);
+	if (*ptr != 'x') {
+		return -EINVAL;
+	}
+	video_mode->height = val;
+
+	ptr++;
+	val = lib_strtoul(ptr, &ptr, 0);
+	if (*ptr != '\0') {
+		return -EINVAL;
+	}
+	video_mode->bpp = val;
+
+	return 0;
+}
+
+
+static void cmd_usage(const char *name)
+{
+	lib_printf(
+		"Usage: %s <options>\n"
+		"\t-p W:H:BPP   preferred mode (default: %dx%dx%d)\n"
+		"\t-f W:H:BPP   fallback mode (default: %dx%dx%d)\n"
+		"\t-h           prints help\n",
+		name, vbe_common.preferred.width, vbe_common.preferred.height, vbe_common.preferred.bpp,
+		vbe_common.fallback.width, vbe_common.fallback.height, vbe_common.fallback.bpp);
+}
+
+
+static void cmd_vbeInfo(void)
+{
+	lib_printf("sets graphics mode using VESA VBE");
+}
+
+
+static int cmd_vbeMain(int argc, char *argv[])
+{
+	int opt;
+	int err;
+
+	lib_printf("\n");
+
+	_vbe_commonInit();
+
+	for (;;) {
+		opt = lib_getopt(argc, argv, "p:f:h");
+		if (opt == -1) {
+			break;
+		}
+
+		if (opt == 'p') {
+			vbe_videoModeParse(optarg, &vbe_common.preferred);
+			continue;
+		}
+
+		if (opt == 'f') {
+			vbe_videoModeParse(optarg, &vbe_common.fallback);
+			continue;
+		}
+
+		cmd_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	do {
+		err = _vbe_infoGet();
+		if (err < 0) {
+			break;
+		}
+
+		graphmode_t graphmode;
```

Comment:
Good practice, but we avoid placing variables anywhere but on the function start in plo and kernel

Reply:
Oops, sorry. Missed it